### PR TITLE
[enhancement](cloud) refine block file cache evict policy

### PR DESCRIPTION
**Pull request by freemandealer**
- Enhance cache performance and reduce cache misses by:
    - Prevent starvation of any particular cache.
    - Improve disk space utilization.
    - Maintain strict and systematic eviction priorities.
- Enhance the observability of caching strategies, allowing cache behavior to be intuitively self-explanatory through monitoring.
- Improve code comprehensibility: unify the cache framework to avoid ad-hoc handling of TTL, increasing the modularity of the code.
- Reduce operational complexity by eliminating and standardizing configuration items.

### CODE REVIEW

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run p0

**Comment by freemandealer** [comment]
run p0

File: be/test/io/cache/block_file_cache_test.cpp [hunk 1]
```diff
@@ -1136,7 +1136,7 @@ TEST_F(BlockFileCacheTest, max_ttl_size) {
         auto holder = cache.get_or_set(key1, offset, 100000, context);
         auto blocks = fromHolder(holder);
         ASSERT_EQ(blocks.size(), 1);
-        if (offset < 90000000) {
+        if (offset < (100000000 * config::max_ttl_cache_ratio / 100)) {
```
Comment (gavinchou) [inline be/test/io/cache/block_file_cache_test.cpp:1139]: better to use `100000000L`

**Comment by gavinchou** [comment]
This PR introduce some behavior change, pls describe what it is

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run cloud_p0

**Comment by freemandealer** [comment]
run buildall

File: be/src/io/cache/block_file_cache.cpp [hunk 2]
```diff
@@ -1203,39 +1313,48 @@ bool BlockFileCache::try_reserve_from_other_queue_by_hot_interval(
...
 }
 
 bool BlockFileCache::try_reserve_from_other_queue_by_size(
-        std::vector<FileCacheType> other_cache_types, size_t size,
+        FileCacheType cur_type, std::vector<FileCacheType> other_cache_types, size_t size,
         std::lock_guard<std::mutex>& cache_lock) {
     size_t removed_size = 0;
     size_t cur_cache_size = _cur_cache_size;
     std::vector<FileBlockCell*> to_evict;
+    // we follow the privilege defined in get_other_cache_types to evict
```
Comment (gavinchou) [inline be/src/io/cache/block_file_cache.cpp:1343]: try to rename `get_other_cache_types` with more meaningful name e.g. `get_other_cache_types_to_evict`

File: be/src/io/cache/block_file_cache.cpp [hunk 3]
```diff
@@ -1579,6 +1700,7 @@ std::string BlockFileCache::reset_capacity(size_t new_capacity) {
                 ss << " ttl_queue released " << queue_released;
             }
             _disk_resource_limit_mode = true;
+            _disk_limit_mode_metrics->set_value(1);
```
Comment (gavinchou) [inline be/src/io/cache/block_file_cache.cpp:1703]: consider setting it to percentage of utilization of the disk if it enterred this mode.
`_disk_limit_mode_metrics` -> `_disk_size_limit_`
and we can also monitor the size utilization and the inode untilization respectively.

File: be/src/io/cache/block_file_cache.cpp [hunk 4]
```diff
@@ -109,14 +197,16 @@ BlockFileCache::BlockFileCache(const std::string& cache_base_path,
                                                            "file_cache_hit_ratio_5m", 0.0);
     _hit_ratio_1h = std::make_shared<bvar::Status<double>>(_cache_base_path.c_str(),
                                                            "file_cache_hit_ratio_1h", 0.0);
+    _disk_limit_mode_metrics =
+            std::make_shared<bvar::Status<size_t>>(_cache_base_path.c_str(), "disk_limit_mode", 0);
```
Comment (gavinchou) [inline be/src/io/cache/block_file_cache.cpp:201]: suggest naming `disk_limit_mode` -> `file_cache_disk_space_limited`
also consider add `file_cache_disk_inode_limited`

File: be/src/io/cache/block_file_cache.cpp [hunk 5]
```diff
@@ -1925,6 +2051,36 @@ std::map<std::string, double> BlockFileCache::get_stats() {
     return stats;
 }
 
+// for be UTs
+std::map<std::string, double> BlockFileCache::get_stats_unsafe() {
```
Comment (gavinchou) [inline be/src/io/cache/block_file_cache.cpp:2055]: private member functions can be access when UT, these code can be put to _test.cpp file

**Review by gavinchou — APPROVED** [review]

**Review by pingchunzhang — APPROVED** [review]

**Pull request merged**